sklearn/datasets/_svmlight_format_fast.pyx

```diff
@@ -113,3 +116,118 @@ def _load_svmlight_file(f, dtype, bint multilabel, bint zero_based,
+    signed long long
+
+ctypedef fused int_or_float2:
+    signed long long
+    double
+
+ctypedef fused int_or_longlong:
+    cython.integral
+    signed long long
+
+def get_dense_row_string(int_or_float1[:,:] X, Py_ssize_t[:] x_inds, int_or_float2[:] x_vals, Py_ssize_t row, str value_pattern, bint one_based):
+    cdef:
+        Py_ssize_t row_length = X.shape[1]
+        Py_ssize_t x_nz_used = 0
+        Py_ssize_t k
+
+    for k in range(row_length):
+        val = X[row,k]
+        if val!=0:
+            x_inds[x_nz_used] = k
+            x_vals[x_nz_used] = val
+            x_nz_used += 1
+    return " ".join(value_pattern % (j+one_based, val) for i, (j, val) in enumerate(zip(x_inds, x_vals)) if i < x_nz_used)
+
+def get_sparse_row_string(int_or_float1[:] X_data, int[:] X_indptr, int[:] X_indices, Py_ssize_t row, str value_pattern, bint one_based):
+    cdef:
+        Py_ssize_t row_start = X_indptr[row]
+        Py_ssize_t row_end = X_indptr[row+1]
+
+    return " ".join(value_pattern % (X_indices[i]+one_based, X_data[i]) for i in range(row_start, row_end))
+
+def _dump_svmlight_file(X, y, f, bint multilabel, bint one_based, int_or_longlong[:] query_id, bint X_is_sp, bint y_is_sp):
+    cdef bint X_is_integral
+    X_is_integral = X.dtype.kind == "i"
+    if X_is_integral:
+        value_pattern = "%d:%d"
+    else:
+        value_pattern = "%d:%.16g"
+    if y.dtype.kind == "i":
+        label_pattern = "%d"
+    else:
+        label_pattern = "%.16g"
+
+    line_pattern = "%s"
+    if query_id is not None:
+        line_pattern += " qid:%d"
+    line_pattern += " %s\n"
+
+    cdef:
+        Py_ssize_t num_labels = y.shape[1]
+        Py_ssize_t x_len = X.shape[0]
+        Py_ssize_t row_length = X.shape[1]
+        Py_ssize_t i
+        Py_ssize_t j
+        Py_ssize_t col_start
+        Py_ssize_t col_end
+        bint first
+        array.array[signed long long] int_template = array.array('q',[])
+        array.array[double] float_template = array.array('d',[])
+        Py_ssize_t x_nz_used
+        Py_ssize_t[:] x_inds = np.empty(row_length, dtype=np.intp)
+        signed long long[:] x_vals_int
+        double[:] x_vals_float
+
+    if not X_is_sp:
+        if X_is_integral:
+            x_vals_int = np.zeros(row_length, dtype=np.longlong)
+        else:
+            x_vals_float = np.zeros(row_length, dtype=np.float64)
+
+    for i in range(x_len):
+        x_nz_used = 0
+
+        if X_is_sp:
+            s = get_sparse_row_string(X.data, X.indptr, X.indices, i, value_pattern, one_based)
+        else:
+            if X_is_integral:
+                s = get_dense_row_string(X, x_inds, x_vals_int, i, value_pattern, one_based)
+            else:
+                s = get_dense_row_string(X, x_inds, x_vals_float, i, value_pattern, one_based)
+        if multilabel:
+            labels_str = ""
+            first = True
+            if y_is_sp:
+                col_start = y.indptr[i]
+                col_end = y.indptr[i+1]
```

Comment:
The code on `main` uses `nonzero`, which detects explicit zeros:

```python
import numpy as np
from scipy.sparse import csr_matrix

indptr = np.array([0, 2, 3, 6])
indices = np.array([0, 2, 2, 0, 1, 2])

# explicit zero in `data`
data = np.array([0, 2, 3, 4, 5, 6])
y = csr_matrix((data, indices, indptr), shape=(3, 3))

There is only one non-zero element at index 2
y.toarray()[0]
# array([0, 0, 2])

# Using nonzero detects only the 2:
y[0].nonzero()[1]
# array([2], dtype=int32)

# using indptr detects a 0 and the 2:
[y.indices[i] for i in range(y.indptr[0], y.indptr[1])]
# [0, 2]
```

This changes the behavior of `dump_svmlight_file`.

Reply:
Added check for explicit zeros and added a test to confirm behaviour as well!

Reply:
With this PR the explicit zeros is serialized, while on `main` the explicit zeros is not serialized. If there are a lot of explicit zeros, then the output of `dump_svmlight_file` will be much bigger compared to `main`.